src/KeyringController.test.ts

```diff
@@ -751,18 +771,121 @@ describe('KeyringController', () => {
       const keyringController = await initializeKeyringController({
         password: PASSWORD,
+        constructorOptions: {
+          encryptor: mockEncryptor,
+        },
       });
       const unsupportedKeyrings = [{ type: 'Ledger Keyring', data: 'DUMMY' }];
-      mockEncryptor.encrypt(PASSWORD, unsupportedKeyrings);
+      await mockEncryptor.encrypt(PASSWORD, unsupportedKeyrings);
       await keyringController.setLocked();
+
       const keyrings = await keyringController.unlockKeyrings(PASSWORD);
+
       expect(keyrings).toHaveLength(0);
       expect(keyringController.unsupportedKeyrings).toStrictEqual(
         unsupportedKeyrings,
       );
     });
+
+    it('should throw error if there is no vault', async () => {
+      const keyringController = new KeyringController({
+        cacheEncryptionKey: false,
+      });
+
+      await expect(async () =>
+        keyringController.unlockKeyrings(PASSWORD),
+      ).rejects.toThrow(KeyringControllerError.VaultError);
+    });
+
+    it('should throw error if decrypted vault is not an array of serialized keyrings', async () => {
+      const mockEncryptor = new MockEncryptor();
+      sinon
+        .stub(mockEncryptor, 'decrypt')
+        .resolves('[{"foo": "not a valid keyring}]');
+      const keyringController = await initializeKeyringController({
+        password: PASSWORD,
+        constructorOptions: {
+          encryptor: mockEncryptor,
+        },
+      });
+
+      await expect(async () =>
+        keyringController.unlockKeyrings(PASSWORD),
+      ).rejects.toThrow(KeyringControllerError.VaultDataError);
+    });
+
+    it('should throw error if decrypted vault includes an invalid keyring', async () => {
+      const mockEncryptor = new MockEncryptor();
+      sinon
+        .stub(mockEncryptor, 'decrypt')
+        .resolves([{ foo: 'not a valid keyring' }]);
+      const keyringController = await initializeKeyringController({
+        password: PASSWORD,
+        constructorOptions: {
+          encryptor: mockEncryptor,
+        },
+      });
+
+      await expect(async () =>
+        keyringController.unlockKeyrings(PASSWORD),
+      ).rejects.toThrow(KeyringControllerError.VaultDataError);
+    });
+
+    describe('with old vault format', () => {
+      describe('with cacheEncryptionKey = true', () => {
+        it('should call persistAllKeyrings', async () => {
+          const updatedVault =
+            '{"vault": "updated_vault_detail", "salt": "salt"}';
+          const mockEncryptor = new MockEncryptor();
+          const updateVaultStub = sinon
+            .stub(mockEncryptor, 'updateVault')
+            .resolves(updatedVault);
+          const keyringController = await initializeKeyringController({
+            password: PASSWORD,
+            constructorOptions: {
+              cacheEncryptionKey: true,
+              encryptor: mockEncryptor,
+            },
+          });
+          const persistAllKeyringsSpy = sinon.spy(
```

Comment:
Nit: Effectively we're testing an internal implementation here by using a spy to check that an internal call is made. Perhaps we could test the effects instead by looking at the store both before and after the keyrings are unlocked?

Reply:
Good point. I refactored this in f2382456a7a4298ffd9df83484ef0fc7345ef23c